### PR TITLE
Adding support for Azure Stack Hub (ASH)

### PR DESCRIPTION
**Pull request by tjungblu**

### CODE REVIEW

File: pkg/operator/starter.go [hunk 1]
```diff
@@ -53,6 +64,31 @@ func RunOperator(ctx context.Context, controllerConfig *controllercmd.Controller
...
+	nodeDaemonSet := "node.yaml"
+
+	runningOnAzureStackHub, err := runningOnAzureStackHub(ctx, configClient.ConfigV1())
+	if err != nil {
+		return err
+	}
+	if runningOnAzureStackHub {
+		controllerDeployment = azureStackCloudFolderPrefix + controllerDeployment
+		volumeSnapshotClass = azureStackCloudFolderPrefix + volumeSnapshotClass
+		nodeDaemonSet = azureStackCloudFolderPrefix + nodeDaemonSet
```
Comment (jsafrane) [inline pkg/operator/starter.go:78]: I personally do not like having two different manifests, they will come out of sync.
In AWS we inject a volume in a Deployment hook like this: https://github.com/openshift/aws-ebs-csi-driver-operator/blob/e8533bf09ce918c008039c15f73cadf98a000546/pkg/operator/starter.go#L211
The same should be possible with a DaemonSet hook.

Comment (tjungblu) [reply pkg/operator/starter.go:78]: sounds good, let me try this out

Comment (tjungblu) [reply pkg/operator/starter.go]: seems to work well with what's there, there's just one exception which is the volumesnapshotclass. There's no hook in the library-go yet. Do you think it makes sense to add another csi controller method for that over there?

File: vendor/modules.txt [hunk 2]
```diff
@@ -90,7 +95,11 @@ github.com/modern-go/concurrent
 github.com/modern-go/reflect2
 # github.com/munnerz/goautoneg v0.0.0-20191010083416-a7dc8b61c822
 github.com/munnerz/goautoneg
+<<<<<<< HEAD
 # github.com/openshift/api v0.0.0-20210521075222-e273a339932a
+=======
+# github.com/openshift/api v0.0.0-20210624153211-ae79113891b0
+>>>>>>> 9735968... add azurestackhub support
 ## explicit
 github.com/openshift/api
```
Comment (gnufied) [inline vendor/modules.txt:104]: conflict?

File: vendor/modules.txt [hunk 3]
```diff
@@ -234,6 +247,12 @@ github.com/spf13/cobra
 # github.com/spf13/pflag v1.0.5
 ## explicit
 github.com/spf13/pflag
+<<<<<<< HEAD
+=======
+# github.com/stretchr/testify v1.6.1
+## explicit
+github.com/stretchr/testify/assert
+>>>>>>> 9735968... add azurestackhub support
```
Comment (gnufied) [inline vendor/modules.txt:255]: same.

File: vendor/modules.txt (hunk 2 above)
Comment (tjungblu) [reply vendor/modules.txt:104]: good catch, let me regenerate the modules

File: pkg/operator/starter.go [hunk 4]
```diff
@@ -62,8 +85,8 @@ func RunOperator(ctx context.Context, controllerConfig *controllercmd.Controller
 		kubeInformersForNamespaces,
 		assets.ReadFile,
 		[]string{
+			volumeSnapshotPath,
 			"storageclass.yaml",
```
Comment (jsafrane) [inline pkg/operator/starter.go:89]: Eventually, we will need a different storage class on AHS (with the default annotation). But it can be a separate PR.

File: assets/storageclass.yaml [hunk 5]
```diff
@@ -2,6 +2,8 @@ apiVersion: storage.k8s.io/v1
 kind: StorageClass
 metadata:
   name: managed-csi
+  annotations:
+    storageclass.kubernetes.io/is-default-class: "true"
```
Comment (jsafrane) [inline assets/storageclass.yaml:6]: On a "regular" Azure Cloud we don't want the CSI storage class to be the default one - we want the in-tree to be the default. You should create a separate StorageClass, as you do with VolumeSnapshotClass

File: pkg/operator/starter.go (hunk 4 above)
Comment (jsafrane) [reply pkg/operator/starter.go:89]: Nevermind, I noticed you update the storage class in this PR. See below, it's bad.

File: assets/storageclass.yaml (hunk 5 above)
Comment (tjungblu) [reply assets/storageclass.yaml:6]: sounds good, I'll split this up.

File: pkg/operator/starter.go (hunk 4 above)
Comment (tjungblu) [reply pkg/operator/starter.go]: done.
